src/agreementManager/AgreementManager.ts

```diff
@@ -206,115 +387,192 @@ class AgreementManager {
+        return agreement?.snapShot;
+    }
+
+    public getLatestStateSnapshot(
+        forkCnt: number
+    ): StateSnapshotStruct | undefined {
+        const agreement = this.forkService.getLatestAgreement(forkCnt);
+        return agreement?.snapShot;
+    }
+
     /**
-     * Gets the latest finalized state (ecnoded) and the latest signed/confirmed state (encoded) from the signer with virtual votes proving it
+     * Gets the latest finalized state (ecnoded) and the latest signed/confirmed state (encoded)
      * @param forkCnt
-     * @param signerAddress
      * @returns
      */
-    public getFinalizedAndLatestWithVotes(
-        forkCnt: BigNumberish,
-        signerAddress: AddressLike
-    ): {
+    public getFinalizedAndLatestWithVotes(forkCnt: BigNumberish): {
         encodedLatestFinalizedState: string;
         encodedLatestCorrectState: string;
-        virtualVotingBlocks: ConfirmedBlockStruct[];
+        virtualVotingBlocks: BlockConfirmationStruct[];
     } {
-        const fork = this.forks.forkAt(Number(forkCnt));
-        if (!fork)
+        const agreementFork = this.forkService.getFork(Number(forkCnt));
+        if (!agreementFork)
             throw new Error(
-                "AgreementManager - getFinalizedAndLatestWithVotes - fork not found"
+                "AgreementManager - getFinalizedAndLatestWithVotes - agreementFork not found"
             );
-        let encodedLatestFinalizedState: string | undefined;
-        let encodedLatestCorrectState: string | undefined;
-        let virtualVotingBlocks: ConfirmedBlockStruct[] = [];
-        let requiredSignatures = SetUtils.fromArray(fork.addressesInThreshold);
 
-        for (const agreement of this.forks.agreementsIterator(
-            forkCnt as number,
-            Direction.BACKWARD
-        )) {
-            const signersAddresses = BlockUtils.getSignerAddresses(
-                agreement.block,
-                agreement.blockSignatures
-            ) as Set<AddressLike>;
+        // iterate over all agreements in the fork from backwards, check if the last agreement has all confirmations,
+        // if not check how many signature are missing and that is the number of virtual votes and the depth of the virtual voting and take the state of that agreement at the depth
+        let encodedLatestFinalizedState = "";
+        let encodedLatestCorrectState = "";
+        let virtualVotingBlocks: BlockConfirmationStruct[] = [];
+        let collectedVotes = 0;
+
+        if (agreementFork.agreements.length === 0) {
+            encodedLatestCorrectState = agreementFork.forkGenesisStateEncoded;
+            encodedLatestFinalizedState = agreementFork.forkGenesisStateEncoded;
+        } else {
+            let totalSignatures =
+                agreementFork.agreements[-1].blockConfirmation.signatures;
+            totalSignatures.push(
+                agreementFork.agreements[-1].blockConfirmation.signedBlock
+                    .signature as BytesLike
```

Comment:
again [-1], but this is less of a problem the problem is that the code logic is not correct - Virtual votes take into account all direct and indirect signatures and they transition to previous blocks recursively and you're looking for a minimal set of linked blocks with signatures (from the author AND others confirming - all the signatures are taken into account) - You don't even check who's signature you're collecting - you just count N. What hurts the most is that you have my code that shows how this is done, docs and even videos explaining and messages in chats and yet again some random stuff is implemented